### DISCUSSION TITLE
Building x11iraf on Mac

### DISCUSSION THREAD

**Discussion by karlglazebrook** [iraf-community/iraf#165] [Q&A] [answered]
Not really an issue - I got these to work.

Note I used the Macports build of X11 so I needed:
```
export C_INCLUDE_PATH=/opt/local/include/
export LIBRARY_PATH=/opt/local/lib

```
before running make. 

There were a few errors at the end of the build:

```
...
cd clients && mkpkg || true
Warning, hlib$mkpkg.inc line 5: macro `IRAFARCH' not found
warning: library `libpkg.a' not found
xc -c -w idxstr.x ism.x wcsgfterm.x
idxstr.c:13:10: fatal error: f2c.h: No such file or directory
   13 | #include "f2c.h"
      |          ^~~~~~~
compilation terminated.
ar r /Users/karl/GIT/x11iraf/ximtool/clients/libpkg.a idxstr.o ism.o wcsgfterm.o
ar: creating archive /Users/karl/GIT/x11iraf/ximtool/clients/libpkg.a
ar: idxstr.o: No such file or directory
ar: ism.o: No such file or directory
ar: wcsgfterm.o: No such file or directory
xc -c -w t_wcspix.x wcimage.x wcmef.x wcmspec.x wcunknown.x
t_wcspix.c:13:10: fatal error: f2c.h: No such file or directory
   13 | #include "f2c.h"
      |          ^~~~~~~
compilation terminated.
ar r /Users/karl/GIT/x11iraf/ximtool/clients/libpkg.a t_wcspix.o wcimage.o wcmef.o wcmspec.o wcunknown.o
ar: t_wcspix.o: No such file or directory
ar: wcimage.o: No such file or directory
ar: wcmef.o: No such file or directory
ar: wcmspec.o: No such file or directory
ar: wcunknown.o: No such file or directory
ranlib libpkg.a
warning: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: empty library: libpkg.a (no table of contents added)
Updated 8 files in libpkg.a
xc -c -w x_ism.x
x_ism.c:13:10: fatal error: f2c.h: No such file or directory
   13 | #include "f2c.h"
      |          ^~~~~~~
compilation terminated.
xc -Nz   -z x_ism.o libpkg.a -o ism_wcspix.e -lds -lxtools -liminterp -lslalib
Warning: file `x_ism.o' not found
ld: file too small (length=8) file 'libpkg.a' for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
Warning, mkpkg line 3: module `relink' not found or returned error
```
though the xgterm and xterm binaries built find and functioned fine. 

I am not sure what f2c is needed for here and I don't have f2c.

**Comment by olebole** [comment, marked as answer]
This part requires to have a working "xc" program from IRAF, and the $iraf environvent variable (and probably $IRAFARCH) properly set.

**Reply by zyuan-astro** [reply]
hello, I have the same error. I can find xc in my iraf path, but wonder how to make things work. Could you give more detailed explanations? Thanks a lot!

**Reply by olebole** [reply]
The "xc" command is part of IRAF, and needs to have the IRAF environment set up, i.e. the correct f2c converter must be installed properly etc. Since I don't know which IRAF version you have, I can't say much about this; for the community version (2.17 is the current one) it is enough to have the `$iraf` environment variable set properly to the IRAF installation where your "xc" (the one that is in your path) belongs to.

So, to debug this further, you need to find out where `$iraf` points to, make sure that this is an IRAF-community version, and then check that `$iraf/unix/f2c/libf2c/f2c.h` exists. also make sure that the `$host` variable is **not** set (or points to `$iraf/unix`).

Again, unless you really need ximtool, there is no need to compile it (and this module). Instead of ximtool, most people nowadays use SAOImageDS9 instead, since it is much more comfortable. In this case, you can just ignore the error.

**Reply by zyuan-astro** [reply]
Hello, Thanks for the reply! I just installed IRAF-2.17 and now try to install x11iraf-2.1

> which xc
> /Users/yuan/.iraf/bin/xc

/Users/yuan/.iraf should be my iraf directory right?

**Reply by olebole** [reply]
No; this is a symbolic link which should point into the IRAF installation directory (usually `$iraf/unix/bin.macos64/xc` or so).
Did you check where `$iraf` points to? And that `$host` is unset?

**Reply by zyuan-astro** [reply]
I see, the xc.e and f2c.e are in the directory` /Users/yuan/iraf-2.17/unix/bin.macos64`

So I should set the following path during the iraf installation, which I did not ... I set the path to be `/Users/yuan/.iraf/bin`

`export PATH=/Users/yuan/iraf-2.17/unix/bin.macos64/:$PATH
`
Is that right?

And how do I find $host?

**Reply by olebole** [reply]
If the correct xc is found, then the `$PATH` is correct. But you need to have `$iraf` set to your IRAF installation, i.e. `/Users/yuan/iraf-2.17/` (trailing slash required!) Is this the case?
If yes, please check that `/Users/yuan/iraf-2.17/unix/f2c/libf2c/f2c.h` exists. 
If `$host` is not there, just ignore it. It is not needed.

**Reply by zyuan-astro** [reply]
Hello! I made it work by adding these lines to .zshrc

> export PATH=/Users/yuan/.iraf/:$PATH
> 

> if [ -e /Users/yuan/.iraf/setup.sh ]; then
>     source /Users/yuan/.iraf/setup.sh
> fi

**Comment by karlglazebrook** [comment]
Yes that seems to be right.

So what does  ism_wcspix.e do ?

Karl

**Comment by olebole** [comment]
It converts pixel coords into world coords, and so allows to have the world coords displayed on the image (for ximtool). If it is not there, ximtool still works but does not display the coordinates.

**Comment by karlglazebrook** [comment]
Ah I see, yes I got that to work too

Karl
